packages/shared/src/hooks/useReadingRank.ts

```diff
@@ -117,30 +117,57 @@ export default function useReadingRank(
       neverShowRankModal,
     );
 
-  const updateShownProgress = async () => {
+  const unmountRef = useRef(false);
+  const timeoutRef = useRef<number>();
+  const visibilityRef = useRef(null);
+
+  const updateShownProgress = useCallback(() => {
+    if (unmountRef) {
+      return;
+    }
+
+    visibilityRef.current = () => {
+      timeoutRef.current = window.setTimeout(updateShownProgress, 1000);
+    };
+
     if (document.visibilityState === 'hidden') {
-      document.addEventListener(
-        'visibilitychange',
-        () => setTimeout(updateShownProgress, 1000),
-        { once: true },
-      );
+      document.addEventListener('visibilitychange', visibilityRef.current, {
+        once: true,
+      });
```

Comment:
I just realized, this condition can be met multiple times, so if there is an existing event handler, I think it should get cleaned up as well. If there is none, this does not result in the app crashing, so this should be safe. What do you think?
```suggestion
      document.removeEventListener('visibilitychange', visibilityRef.current);
      document.addEventListener('visibilitychange', visibilityRef.current, {
        once: true,
      });
```

Reply:
Yep 100% agree, we missed that one!